pkg/media/webrtc_sink.go

```diff
@@ -32,43 +34,98 @@ import (
 )
 
 type WebRTCSink struct {
-	params *params.Params
+	params    *params.Params
+	onFailure func()
+
+	lock     sync.Mutex
+	sdkReady core.Fuse
+	closed   core.Fuse
+	errChan  chan error
 
 	sdkOut        *lksdk_output.LKSDKOutput
 	outputSync    *utils.OutputSynchronizer
 	statsGatherer *stats.LocalMediaStatsGatherer
 }
 
-func NewWebRTCSink(ctx context.Context, p *params.Params, statsGatherer *stats.LocalMediaStatsGatherer) (*WebRTCSink, error) {
+func NewWebRTCSink(ctx context.Context, p *params.Params, onFailure func(), statsGatherer *stats.LocalMediaStatsGatherer) (*WebRTCSink, error) {
 	ctx, span := tracer.Start(ctx, "media.NewWebRTCSink")
 	defer span.End()
 
-	sdkOut, err := lksdk_output.NewLKSDKOutput(ctx, p)
-	if err != nil {
-		return nil, err
-	}
-
-	return &WebRTCSink{
+	s := &WebRTCSink{
 		params:        p,
-		sdkOut:        sdkOut,
+		onFailure:     onFailure,
+		errChan:       make(chan error),
 		outputSync:    utils.NewOutputSynchronizer(),
 		statsGatherer: statsGatherer,
-	}, nil
+	}
+
+	go func() {
+		var err error
+
+		defer func() {
+			s.sdkReady.Break()
+			if err != nil && s.onFailure != nil {
+				select {
+				case s.errChan <- err:
+				default:
+				}
+				s.onFailure()
+			}
+		}()
+
+		sdkOut, err := lksdk_output.NewLKSDKOutput(ctx, p)
+		if err != nil {
+			return
+		}
+
+		s.lock.Lock()
+		s.sdkOut = sdkOut
+		s.lock.Unlock()
+	}()
+
+	return s, nil
 }
 
 func (s *WebRTCSink) addAudioTrack() (*Output, error) {
-	track, err := s.sdkOut.AddAudioTrack(putils.GetMimeTypeForAudioCodec(s.params.AudioEncodingOptions.AudioCodec), s.params.AudioEncodingOptions.DisableDtx, s.params.AudioEncodingOptions.Channels > 1)
-	if err != nil {
-		return nil, err
-	}
-
-	output, err := NewAudioOutput(s.params.AudioEncodingOptions, s.outputSync.AddTrack(), track, s.statsGatherer)
+	output, err := NewAudioOutput(s.params.AudioEncodingOptions, s.outputSync.AddTrack(), s.statsGatherer)
 	if err != nil {
 		logger.Errorw("could not create output", err)
 		return nil, err
 	}
 
-	s.sdkOut.AddOutputs(output)
+	go func() {
+		var sdkOut *lksdk_output.LKSDKOutput
+		var err error
+
+		defer func() {
+			if err != nil && s.onFailure != nil {
+				select {
+				case s.errChan <- err:
+				default:
+				}
+				s.onFailure()
+			}
+		}()
+
+		select {
+		case <-s.closed.Watch():
+		case <-s.sdkReady.Watch():
+			s.lock.Lock()
+			sdkOut = s.sdkOut
+			s.lock.Unlock()
+		}
+
+		if sdkOut != nil {
+			track, err := sdkOut.AddAudioTrack(putils.GetMimeTypeForAudioCodec(s.params.AudioEncodingOptions.AudioCodec), s.params.AudioEncodingOptions.DisableDtx, s.params.AudioEncodingOptions.Channels > 1)
```

Comment:
The `err` here is likely shadowed and is not the same that is checked in `defer`. You'd have to define `var track` too.

Reply:
Good catch, thanks!